sycl/doc/extensions/proposed/sycl_ext_oneapi_execution_domain_info.asciidoc

```diff
@@ -0,0 +1,242 @@
+= sycl_ext_oneapi_execution_domain_info
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+:endnote: &#8212;{nbsp}end{nbsp}note
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2024 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 9 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+SYCL 2020 allows developers to query the number of _compute units_ in a device,
+where each compute unit describes a place where work-groups can execute.
+However, the definition of compute unit is sufficiently vague that different
+implementations and backends interpret it differently, which makes it difficult
+for developers to use the number of compute units in a portable way.
+
+This extension introduces the concept of an _execution domain_ and associated
+device queries to SYCL in order to address this portability issue.
+The definition of execution domain proposed here is intended to be more precise
+than the definition of compute unit -- to provide users with certain guarantees
```

Comment:
It wasn't clear to me from the description below what guarantee is being provided.  It seems to me that the new query is more expressive than "compute units" because it provides a separate value for each execution scope.  However, the description of the value that is returned for each scope seems just as vague as "compute units".  Am I missing something?

Reply:
My idea here was to add extra constraints via the definition of "execution domain", rather than via the query definitions.  Currently there's only one real constraint, which is:

> A given group of SYCL work-items with execution scope _S_ must execute in exactly one execution domain associated with the same execution scope.

What I was trying to do here was prevent the situation we have today, where some implementations use multiple compute units to execute a single work-group.  But looking at OpenCL again more closely, I see that it defines:

> **Compute Unit**: An OpenCL device has one or more compute units. A work-group executes on a 
single compute unit. A compute unit is composed of one or more processing elements and local 
memory. A compute unit may also include dedicated texture filter units that can be accessed by 
its processing elements. 

SYCL 2020 doesn't include a definition of "Compute Unit" but it does say:

> The work-items in a given work-group execute on the processing elements of a single compute unit.

...so I might need to think about this a bit more.  I'm going to convert this to a draft while I do that.